File: backends/ts-hono/src/routes/staffs.ts
/**
 * スタッフルーターモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
import { Hono } from "hono";
import { badRequest } from "../lib/errors.js";
import { formatTime, getStaffIdFromCookie } from "../lib/cookie.js";
import { buildPager } from "../lib/pager.js";
import { db, asTx } from "../db/client.js";
import { DrizzleStaffRepository } from "../infrastructure/persistence/drizzleStaffRepository.js";
import { StaffInteractor } from "../usecase/staff/interactor.js";
import type { StaffListItem } from "../domain/staff/valueObjects.js";

const app = new Hono();

function mapStaff(s: StaffListItem) {
  return {
    id: s.id, name: s.name, email: s.email, role: s.role, status: s.status,
    created_at: formatTime(s.createdAt), updated_at: formatTime(s.updatedAt),
  };
}

app.get("/staffs", async (c) => {
  const keyword = c.req.query("keyword");
  const rolesRaw = c.req.queries("roles") ?? [];
  const roles = rolesRaw.flatMap(r => r.split(",")).map(Number).filter(n => !isNaN(n));
  const limit = Math.max(1, parseInt(c.req.query("limit") ?? "20", 10) || 20);
  const page = Math.max(1, parseInt(c.req.query("page") ?? "1", 10) || 1);
  const offset = limit * (page - 1);
  const sort = c.req.query("sort");
  const sortType = c.req.query("sort_type");

  const uc = new StaffInteractor(new DrizzleStaffRepository(db));
  const [list, count] = await uc.findByCondition(keyword, roles, offset, limit, sort, sortType);
  const pager = buildPager(count, limit, offset, list.length);
  return c.json({ data: list.map(mapStaff), pager });
});

app.patch("/staffs/:id/updateRole", async (c) => {
  const id = parseInt(c.req.param("id"), 10);
  const body = await c.req.json<{ role?: number }>();
  if (body.role === undefined) throw badRequest("role_required");
  const executorId = getStaffIdFromCookie(c);
  await db.transaction(async (tx) => {
    const uc = new StaffInteractor(new DrizzleStaffRepository(asTx(tx)));
    await uc.updateRole(id, body.role!, executorId);
  });
  return c.json({ id });
});

app.patch("/staffs/:id/restore", async (c) => {
  const id = parseInt(c.req.param("id"), 10);
  await db.transaction(async (tx) => {
    const uc = new StaffInteractor(new DrizzleStaffRepository(asTx(tx)));
    await uc.restore(id);
  });
  return c.json({ id });
});

app.delete("/staffs/:id/delete", async (c) => {
  const id = parseInt(c.req.param("id"), 10);
  const executorId = getStaffIdFromCookie(c);
  await db.transaction(async (tx) => {
    const uc = new StaffInteractor(new DrizzleStaffRepository(asTx(tx)));
    await uc.destroy(id, executorId);
  });
  return c.json({ id });
});

export default app;

File: backends/python-fastapi/app/routers/staffs.py
"""
スタッフルーターモジュール。

Author: Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
"""
import math
from typing import Optional
from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from app.routers.deps import get_staff_interactor, get_staff_id_from_cookie
from app.usecase.staff.interactor import StaffInteractor
from app.usecase.staff.dto import StaffUpdateRoleDto, StaffDestroyDto

router = APIRouter()

_DEFAULT_PAGE_COUNT = 5


def _build_pager(count: int, limit: int, offset: int, record_count: int) -> dict:
    if limit <= 0:
        limit = 20
    page_count = max(1, math.ceil(count / limit))
    last_page_offset = (page_count * limit) - limit
    if count > 0 and offset > last_page_offset:
        offset = last_page_offset
    page = int(math.floor(math.ceil(offset / limit))) + 1 if limit > 0 else 1
    start_page = max(1, page - (_DEFAULT_PAGE_COUNT - 1))
    end_page = min(page_count, start_page + (_DEFAULT_PAGE_COUNT - 1))
    return {
        "count": count,
        "limit": limit,
        "next": page_count > page,
        "previous": page > 1,
        "page": page,
        "nextPage": page + 1,
        "previousPage": page - 1,
        "pageCount": page_count,
        "first": page > 1,
        "last": page_count > page,
        "firstRecordCount": offset + 1,
        "lastRecordCount": offset + record_count,
        "startPage": start_page,
        "endPage": end_page,
    }


def _map_staff(s) -> dict:
    return {
        "id": s.id,
        "name": s.name,
        "email": s.email,
        "role": s.role,
        "status": s.status,
        "created_at": s.created_at.strftime("%Y-%m-%d %H:%M") if s.created_at else None,
        "updated_at": s.updated_at.strftime("%Y-%m-%d %H:%M") if s.updated_at else None,
    }


@router.get("/staffs")
def index(
    keyword: Optional[str] = Query(default=None),
    roles: Optional[list[int]] = Query(default=None),
    limit: int = Query(default=20, ge=1),
    page: int = Query(default=1, ge=1),
    sort: Optional[str] = Query(default=None),
    sort_type: Optional[str] = Query(default=None),
    interactor: StaffInteractor = Depends(get_staff_interactor),
):
    offset = limit * (page - 1)
    staffs, count = interactor.find_by_condition(
        keyword=keyword, roles=roles or [], offset=offset, limit=limit, sort=sort, sort_type=sort_type
    )
    pager = _build_pager(count, limit, offset, len(staffs))
    return {"data": [_map_staff(s) for s in staffs], "pager": pager}


class UpdateRoleBody(BaseModel):
    role: int


@router.patch("/staffs/{staff_id}/updateRole")
def update_role(
    staff_id: int,
    body: UpdateRoleBody,
    executor_id: int = Depends(get_staff_id_from_cookie),
    interactor: StaffInteractor = Depends(get_staff_interactor),
):
    dto = StaffUpdateRoleDto(staff_id=staff_id, role=body.role, executor_id=executor_id)
    interactor.update_role(dto)
    return {"id": staff_id}


@router.patch("/staffs/{staff_id}/restore")
def restore(staff_id: int, interactor: StaffInteractor = Depends(get_staff_interactor)):
    interactor.restore(staff_id)
    return {"id": staff_id}


@router.delete("/staffs/{staff_id}/delete")
def destroy(
    staff_id: int,
    executor_id: int = Depends(get_staff_id_from_cookie),
    interactor: StaffInteractor = Depends(get_staff_interactor),
):
    dto = StaffDestroyDto(staff_id=staff_id, executor_id=executor_id)
    interactor.destroy(dto)
    return {"id": staff_id}
